torch/_decomp/decompositions.py

```diff
@@ -3660,6 +3659,131 @@ def get_values(inp_size, out_size, scales, nsqueeze):
+    input_size: List[int],
+    align_corners: bool,
+    scales_h: Optional[float] = None,
+    scales_w: Optional[float] = None,
+) -> Tensor:
+    return _upsample_linear_backward(
+        grad_output, output_size, input_size, align_corners, [scales_h, scales_w]
+    )
+
+
+@register_decomposition(aten.upsample_trilinear3d_backward)
+@out_wrapper("grad_input")
+def upsample_trilinear3d_backward(
+    grad_output: Tensor,
+    output_size: List[int],
+    input_size: List[int],
+    align_corners: bool,
+    scales_d: Optional[float] = None,
+    scales_h: Optional[float] = None,
+    scales_w: Optional[float] = None,
+) -> Tensor:
+    return _upsample_linear_backward(
+        grad_output,
+        output_size,
+        input_size,
+        align_corners,
+        [scales_d, scales_h, scales_w],
+    )
+
+
+@pw_cast_for_opmath
+def _upsample_linear_backward(
+    grad_output: Tensor,
+    output_size: List[int],
+    input_size: List[int],
+    align_corners: bool,
+    scales: List[Optional[float]],
+) -> Tensor:
+    # get dimensions of original image
+    n_batch, n_channels = input_size[:2]
+    inp_sizes = input_size[2:]
+    n_dims = len(inp_sizes)
+
+    _, dtype = utils.elementwise_dtypes(
+        grad_output,
+        type_promotion_kind=utils.ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT,
+    )
+
+    def get_values(inp_size, out_size, scales, nsqueeze):
+        # First Calculate scaling factor
+        scale_factor = _compute_scale(inp_size, out_size, align_corners, scales)
+        # We have to create arange with int64 dtype and use .to in order to avoid
+        # additional kernels creation in inductor and get a perf slowdown
+        i = torch.arange(out_size, device=grad_output.device).to(dtype=dtype)
+
+        x_f32 = _compute_source_index(scale_factor, i, align_corners).clamp(min=0.0)
+        x_f32 = x_f32.reshape(x_f32.shape[0], *[1] * (nsqueeze))
+        x = x_f32.to(torch.int64)
+        xp1 = (x + 1).clamp(max=inp_size - 1)
+        return x_f32, x, xp1
+
+    values = [
+        get_values(inp_size, out_size, scales, n_dims - 1 - i)
+        for i, (inp_size, out_size, scales) in enumerate(
+            zip(inp_sizes, output_size, scales)
+        )
+    ]
+    xs_f32, xs, xp1s = list(zip(*values))
+
+    # Using functions here to help inductor do fusions easier.
+    # This results in (a * (b * (c * x)) instead of (a * (b * c)) * x
+    # in the generated code and therefore inductor does not try to
+    # create temporaries for (b * c).
+    coeff_fns = [lambda x: x]
+    for i in range(n_dims):
+        xscale = (xs_f32[i] - xs[i]).clamp(0.0, 1.0).to(dtype)
+        new_coeff_fns: List[Any] = [None] * (2 * len(coeff_fns))
+        new_coeff_fns[::2] = [
+            lambda x: torch.mul(coeff_fn(x), (1 - xscale)) for coeff_fn in coeff_fns
+        ]
+        new_coeff_fns[1::2] = [
+            lambda x: torch.mul(coeff_fn(x), xscale) for coeff_fn in coeff_fns
+        ]
+        coeff_fns = new_coeff_fns
+
```

Comment:
would you mind filing a standalone issue here on the optimization inductor should be doing ?

Reply:
Found https://github.com/pytorch/pytorch/issues/124653 while trying to create a MWE. Not sure if it's the same bug.